### PR TITLE
Concurrent drains

### PR DESCRIPTION
**Pull request by asvetlov**
asyncio doesn't support concurrent `.drain()` calls (https://github.com/aio-libs/aiohttp/pull/1886#issuecomment-301355495)
This PR proposes a workaround for the issue.

### CODE REVIEW

File: aiohttp/http_writer.py [hunk 1]
```diff
@@ -125,7 +128,18 @@ def drain(self):
           yield from w.drain()
         """
         if self._protocol.transport is not None:
-            yield from self._protocol._drain_helper()
+            if self._is_drain:
+                fut = asyncio.Future(loop=self._loop)
```
Comment (popravich) [inline aiohttp/http_writer.py:132]: `self._loop.create_future()` I guess

File: aiohttp/http_writer.py [hunk 2]
```diff
@@ -125,7 +128,18 @@ def drain(self):
...
+                self._drain_waiters.append(fut)
+                yield from fut
+            self._is_drain = True
+            try:
+                yield from self._protocol._drain_helper()
+            finally:
+                self._is_drain = False
+                if self._drain_waiters:
+                    fut = self._drain_waiters.popleft()
+                    fut.set_result(None)
```
Comment (popravich) [inline aiohttp/http_writer.py:142]: The same "deadlock" issue (a,b) can happen here: if future on line 134 is cancelled then other task that
sets result for next waiter will do nothing and all other futures will be in `_drain_waiters` forever.

a: bugs.python.org/issue27585
b: https://github.com/python/asyncio/pull/393 and propbably https://github.com/python/cpython/pull/1031

Comment (popravich) [reply aiohttp/http_writer.py:142]: I think cancellation tests should be added as well.

File: aiohttp/http_writer.py (hunk 1 above)
Comment (fafhrd91) [inline aiohttp/http_writer.py:132]: can't we use same future for all callers?

**Comment by asvetlov** [comment]
The code is outdated

**Pull request closed without merging**